From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Thu, 3 Mar 2022 12:36:17 +0700
Subject: [PATCH] Jettpack: randomTick blocks below y=0 and above y=255

This patch was originally created under the GNU Lesser General Public License v3.0 for the "JettPack" project: https://gitlab.com/Titaniumtown/JettPack

diff --git a/src/main/java/com/destroystokyo/paper/util/maplist/IBlockDataList.java b/src/main/java/com/destroystokyo/paper/util/maplist/IBlockDataList.java
index afd194c5e63f4f8e2220c5b382138e96ecb14c1b..7472b9f4e0bd855f169680db8c90f738e7a2ffd2 100644
--- a/src/main/java/com/destroystokyo/paper/util/maplist/IBlockDataList.java
+++ b/src/main/java/com/destroystokyo/paper/util/maplist/IBlockDataList.java
@@ -24,7 +24,7 @@ public final class IBlockDataList {
     private int size;
 
     public static int getLocationKey(final int x, final int y, final int z) {
-        return (x & 15) | (((z & 15) << 4)) | ((y & 255) << (4 + 4));
+        return (x & 15) | ((z & 15) << 4) | ((y & 4095) << 8); // JettPack
     }
 
     public static BlockState getBlockDataFromRaw(final long raw) {
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index b1ea6826794f8187a88ff99671bdafcb00095164..9120edc92c30233bc078501225b4f9ad95a3d020 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -881,7 +881,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                     continue;
                 }
 
-                int yPos = (sectionIndex + minSection) << 4;
+                int yPos = ((sectionIndex + minSection) << 4) + 4096; // Jettpack
                 for (int a = 0; a < randomTickSpeed; ++a) {
                     int tickingBlocks = section.tickingList.size();
                     int index = this.randomTickRandom.nextInt(16 * 16 * 16);
@@ -892,7 +892,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                     long raw = section.tickingList.getRaw(index);
                     int location = com.destroystokyo.paper.util.maplist.IBlockDataList.getLocationFromRaw(raw);
                     int randomX = location & 15;
-                    int randomY = ((location >>> (4 + 4)) & 255) | yPos;
+                    int randomY = (((location >>> 8) & 4095) | yPos) - 4096; // Jettpack
                     int randomZ = (location >>> 4) & 15;
 
                     BlockPos blockposition2 = blockposition.set(j + randomX, randomY, k + randomZ);
diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
index d69fe40d3af983144999963ec25b32500237e515..d25784481496ab7f6f6d745f140a6de673b5e7eb 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -190,7 +190,7 @@ public class LevelChunkSection {
             if (iblockdata1.isRandomlyTicking()) {
                 --this.tickingBlockCount;
                 // Paper start
-                this.tickingList.remove(x, y, z);
+                this.tickingList.remove(x, y + 4096, z); // Jettpack
                 // Paper end
             }
         }
@@ -205,7 +205,7 @@ public class LevelChunkSection {
             if (state.isRandomlyTicking()) {
                 ++this.tickingBlockCount;
                 // Paper start
-                this.tickingList.add(x, y, z, state);
+                this.tickingList.add(x, y + 4096, z, state); // Jettpack
                 // Paper end
             }
         }
